python/ros_wrapper/helpers/viz.py

```diff
@@ -193,10 +218,99 @@ def gt_cb_3(self,msg):
+
+        # build list of distance one and distance two neighbors for each agent
+        # each agent gets full list of connections
+        neighbor_conn_ids = []
+        for j in range(0,len(self.connections[id_index])):
+            for k in range(0,len(self.connections[self.ordered_ids.index(self.connections[id_index][j])])):
+
+                if not self.connections[self.ordered_ids.index(self.connections[id_index][j])][k] in neighbor_conn_ids:
+                    neighbor_conn_ids += self.connections[self.ordered_ids.index(self.connections[id_index][j])]
+
+                # remove agent's own id from list of neighbors
+                if agent_id in neighbor_conn_ids:
+                    neighbor_conn_ids.remove(agent_id)
+
+        # combine with direct connection ids and sort
+        conn_ids = list(set(sorted(conn_ids + neighbor_conn_ids)))
+
+        # divide out direct measurement connections and all connections
+        connections_new = list(set(sorted(neighbor_conn_ids + self.connections[id_index])))
+
+        # find ownship location in ids
+        ownship_index = conn_ids.index(agent_id)
+
+        estimate_mean = np.take(msg.mean[6*ownship_index:6*ownship_index+6],[0,2,4])
+        estimate_covariance = inflate_covariance(msg.covariance)[6*ownship_index:6*ownship_index+6][np.ix_([0,2,4],[0,2,4])]
+
+        # compute covariance ellipse scaling and rotation (quaternion comes out as [w,x,y,z])
+        scale, rotation_quat = uncertain_ellipse(estimate_covariance)
+
+        marker_color_list = MARKER_COLOR[id_index]
+
+        new_marker = Marker()
+        new_marker.header.frame_id = "map"
+        new_marker.header.stamp = rospy.Time()
+        new_marker.id = msg.src
+        new_marker.type = Marker.SPHERE # SPHERE
+        new_marker.action = Marker.ADD # MODIFY
+        new_marker.pose.position.x = estimate_mean[0]
+        new_marker.pose.position.y = estimate_mean[1]
+        new_marker.pose.position.z = estimate_mean[2]
+        new_marker.pose.orientation.x = rotation_quat[1]
+        new_marker.pose.orientation.y = rotation_quat[2]
+        new_marker.pose.orientation.z = rotation_quat[3]
+        new_marker.pose.orientation.w = rotation_quat[0]
+        new_marker.scale.x = scale[0]
+        new_marker.scale.y = scale[1]
+        new_marker.scale.z = scale[2]
+        new_marker.color.a = 0.5
+        new_marker.color.r = marker_color_list[0]
+        new_marker.color.g = marker_color_list[1]
+        new_marker.color.b = marker_color_list[2]
+        new_marker.lifetime = rospy.Duration(100)
+
+        self.marker_pub.publish(new_marker)
+
+def uncertain_ellipse(covariance,alpha=0.95):
+    """
+    Compute semi-major and semi-minor axis dimensions, and rotation.
+    Alpha defines confidence level for ellipse.
+    Assumes 3x3 covariance.
+    """
+    # compute eigenvalues and vectors of covariance
+    l, v = np.linalg.eig(covariance)
```

Comment:
Need to ensure largest to smallest ordering of eigenvalues, and apply that ordering to the eigenvectors.

Reply:
This will be addressed in a subsequent commit for viz.py after merge